dips/dip-16.md

```diff
@@ -0,0 +1,120 @@
+status: Draft
+type: Informational
+created: 3/24/2021
+---
+
+Diem Framework transactions contain Move transaction scripts that change on-chain state (e.g., balances) by invoking functions of Move modules published on-chain. This DIP explains both the mechanics of transaction scripts and the process for adding, removing, and changing scripts. This DIP does *not* document the specific scripts that are available to Diem users, but the complete list of scripts can be found in the [Diem developer docs](https://github.com/diem/diem/blob/main/language/diem-framework/transaction_scripts/doc/transaction_script_documentation.md).
+
+### Transaction payloads
+
+A Diem transaction includes a `TransactionPayload` that contains the logic for performing an on-chain update:
+
+```
+pub enum TransactionPayload {
+    /// A system maintenance transaction.
+    WriteSet(WriteSetPayload),
+    /// A transaction that executes code.
+    Script(Script),
+    /// A transaction that publishes code.
+    Module(Module),
+    /// A transaction that executes an existing script function published on-chain.
+    ScriptFunction(ScriptFunction),
+}
+```
+
+The two `TransactionPayload` variants relevant to this DIP are `ScriptFunction` (which contains a pointer to a function already published in an on-chain Move module) and `Script` (a standalone, single-function bytecode program).
+
+## Script signature requirements
+
+Both script functions and the single function in a transaction script bytecode file have the following restrictions. For a function` f<ability_params>(param_types): ret_types` :
+
+* The `ret_types` list is empty (i.e., the function does not return a value)
+* The `param_types` list begins with one or more `signer` types
+* No `signer` type appears after a non-`signer` type in `param_types`
+    * E.g., `f(signer)`, `f(signer, u64)`, and `f(signer, signer, bool)` satisfy this condition
+    * E.g., `f(u64, signer)`, `f(signer, u64, signer)`, `f(signer, bool, u64, signer)` do not satisfy this condition
+* Each non-`signer` type in `param_types` has the `copy` ability and is not a struct
+    * E.g., `f(signer, Diem<XUS>`) does not satisfy this condition because `Diem<XUS>` is a struct
+* Each type in `param_types` is *closed* (that is, does not refer to the type variables in `ability_params`
+    * E.g., `f<T>(signer, vector<T>)` does not satisfy this condition
+    * E.g., `f(signer, vector<u8>`) does satisfy this condition
+
+## Script Functions
+
+In the Move language, a function can be declared with four different visibility levels: `public`, `private`, `public(friend)`, and `public(script)`. *Any* function with `public(script)` visibility can be invoked by a transaction if it satisfies the script signature requirements explained below.
+
+A client can define a `ScriptFunction` transaction payload using the following struct:
+
+```
+pub struct ScriptFunction {
+    module: ModuleId,
+    function: Identifier,
+    ty_args: Vec<TypeTag>,
```

Comment:
are these defined anywhere? can we point to them?

Reply:
They are all defined in move-core-types. I'll add a note below.